File: src/main/java/com/demonwav/mcdev/update/PluginUpdateStatus.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2016 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.update;

import com.intellij.ide.plugins.IdeaPluginDescriptor;
import com.intellij.util.text.VersionComparatorUtil;

import java.io.PrintWriter;
import java.io.StringWriter;

public class PluginUpdateStatus {

    private PluginUpdateStatus() {}

    public PluginUpdateStatus mergeWith(PluginUpdateStatus other) {
        // Jesus wept. https://github.com/JetBrains/kotlin/blob/master/idea/src/org/jetbrains/kotlin/idea/KotlinPluginUpdater.kt#L61-L63
        if (
            other instanceof Update &&
                (this instanceof LatestVersionInstalled ||
                    (this instanceof Update && VersionComparatorUtil.compare(
                        ((Update) other).getPluginDescriptor().getVersion(),
                        ((Update) this).getPluginDescriptor().getVersion()) > 0
                    )
                )
        ) {
            return other;
        }
        return this;
    }

    public static class LatestVersionInstalled extends PluginUpdateStatus {}

    public static PluginUpdateStatus fromException(String message, Exception e) {
        final StringWriter stringWriter = new StringWriter();
        e.printStackTrace(new PrintWriter(stringWriter));
        return new CheckFailed(message, stringWriter.toString());
    }

    public static class Update extends PluginUpdateStatus {
        final IdeaPluginDescriptor pluginDescriptor;
        final String hostToInstallFrom;

        public Update(IdeaPluginDescriptor pluginDescriptor, String hostToInstallFrom) {
            this.pluginDescriptor = pluginDescriptor;
            this.hostToInstallFrom = hostToInstallFrom;
        }

        public IdeaPluginDescriptor getPluginDescriptor() {
            return pluginDescriptor;
        }

        public String getHostToInstallFrom() {
            return hostToInstallFrom;
        }
    }

    public static class CheckFailed extends PluginUpdateStatus {
        final String message;
        final String detail;

        public CheckFailed(String message, String detail) {
            this.message = message;
            this.detail = detail;
        }

        public String getMessage() {
            return message;
        }

        public String getDetail() {
            return detail;
        }
    }
}

File: src/main/kotlin/com/demonwav/mcdev/update/PluginUpdateStatus.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.update

import com.intellij.ide.plugins.IdeaPluginDescriptor
import com.intellij.util.text.VersionComparatorUtil

sealed class PluginUpdateStatus : Comparable<PluginUpdateStatus> {

    protected abstract fun getVersionString(): String?

    fun mergeWith(other: PluginUpdateStatus): PluginUpdateStatus = maxOf(this, other)

    override fun compareTo(other: PluginUpdateStatus): Int {
        val thisVersion = this.getVersionString() ?: return 1
        val otherVersion = other.getVersionString() ?: return -1
        return VersionComparatorUtil.compare(thisVersion, otherVersion)
    }

    object LatestVersionInstalled : PluginUpdateStatus() {
        override fun getVersionString(): String? = PluginUtil.pluginVersion
    }

    class Update(val pluginDescriptor: IdeaPluginDescriptor, val hostToInstallFrom: String?) : PluginUpdateStatus() {
        override fun getVersionString(): String? = this.pluginDescriptor.version
    }

    class CheckFailed(val message: String) : PluginUpdateStatus() {
        override fun getVersionString(): String? = null
    }
}
